### DISCUSSION TITLE
Note on the state of Nelua

### DISCUSSION THREAD

**Discussion by bentxt** [edubart/nelua-lang#222] [General]
Hi,
While playing with the compiler and the examples, while reading documentation and some of the discussions here I get the impression that Nelua is a solid project. From what I learn the author works on the language since a few years continuously and seem to have a pretty firm understand of what Nelua provides and how it should work. So I'm a little confused how to read the note on the frontpage: "The language is in alpha state and still evolving."

I already admitted that my knowledge is really superficial. And I ran into very unfinished stuff, like the lua generator. But the author explains that somewhere here in the discussion section. So far I didn't saw anything totally unexpected and the numerous discussions here never made the impression that this is an immature project and the way of learning its first hard lessons. It all seems pretty mature and conceptually finished.

So my question is: are there any fundamental holes, key functionalities that are missing. Do you have any kind of roadmap on how to leave the alpha state?

And finally and in reference to the original title. In my opinion, you should may consider to word that alpha note differently. I find it a little misleading and not mirroring the state of the project.

Thanks for all the great work
Ben

**Comment by edubart** [comment]
> So my question is: are there any fundamental holes, key functionalities that are missing. Do you have any kind of roadmap on how to leave the alpha state?

Big language features missing:
- No builtin sumtypes, they are planned but not implemented yet. (however you can implement them using union).
- No builtin optional types, they are planned but not implemented yet. (however you can implement them using records).
- No builtin any runtime type, they are planned but not implemented yet. (you have to use static typing instead).
- No dynamic tables like in Lua, they are planned but not implemented yet. (you have to use hashmaps or vectors instead).
- No closures like in Lua, they are planned but not implemented yet. (you have to workaround with opaque pointers like in C).
- No exception system like in Lua pcall, they are planned but not implemented yet. 

If you come from Lua or other high level language and see the above missing features you may be disappointed. However if you come from the C systems programming language, you will not bother. So I guess the answer to that question depends where you came from, and what you want. 

Also other things not finished:
- Allocator system for manual memory management is not finished.
- Automatic C binding generator is not polished or documented.
- There is a plan to implement a "safe" strict subset of the language so newcomers can safely use the language without shooting their foot with memory related crashes, it's not there yet.
- Debugging you have to use GDB and read C code, but not much of an issue because the C code is very readable, but people not wanting to touch C may get disappointed.

So many planned features are missing, but hey there are many people using it despite that. Nelua can completely replace C with what it has so far, but we can't say the same for Lua, which was its long term goal. For the features Nelua has so far, it feels stable.

> Do you have any kind of roadmap on how to leave the alpha state?

My roadmap would be all the features I mentioned above to leave alpha state, plus some other cleanups. Then I could declare the language "finished". But maybe that is too ambitious. All those features are hard to implement, but possible, and currently I have limited time to develop them, so don't keep waiting for them.

**Reply by stefanos82** [reply]
@edubart Hmm, maybe renaming `alpha state` to `bleeding edge state` could make it more obvious what development model is currently following? 

This way it could help clear things up for newcomers to know they are dealing with a project that evolves continuously.

P.S.: [Off-topic of current thread:] I have noticed an incredible activity by you around garbage collection; coincidentally, 6 days ago the second edition of [The Garbage Collection Handbook, 2nd Edition](https://www.routledge.com/The-Garbage-Collection-Handbook-The-Art-of-Automatic-Memory-Management/Jones-Hosking-Moss/p/book/9781032218038) got released. Did you buy it and got inspired or it's just sheer coincidence?

**Reply by edubart** [reply]
> Have noticed an incredible activity by you around garbage collection; coincidentally, 6 days ago the second edition of [The Garbage Collection Handbook, 2nd Edition](https://www.routledge.com/The-Garbage-Collection-Handbook-The-Art-of-Automatic-Memory-Management/Jones-Hosking-Moss/p/book/9781032218038) got released.

This was just a coincidence, but thanks for the reference. I was improving the GC because I started to think about a "safe" Nelua subset, so newcomers could use it to make programs without foot guns and crashes. In the central idea of this "safe" subset the GC, turn the GC off you are really on your own and you have all foot guns like C has, and I thought if I want to encourage newcomers to use the GC, I might as well revisit it and optimize it, so my recent commits made some cleanups and optimizations, and the GC is about 1.6x faster.

**Reply by stefanos82** [reply]
I thought it was my idea, but now you have confirmed my initial suspicion; indeed it is faster now, because I have noticed the tremendous difference in execution, even though the compilation also got improved quite a lot, which probably is unrelated to GC optimizations.

**Comment by bentxt** [comment]
Thanks for the great answers. I have some follow questions. 
But why is this thread already closed?  I agree that it is 'answered', but maybe this discussion is also interesting for others.
So could you leave it open?

**Reply by RaycatWhoDat** [reply]
He probably closed it because your original questions were answered. A discussion marked as "closed" doesn't mean other people can't see it. But if you want to continue your other topic as a long-running discussion, you'd do better by making a new discussion separate from this one and referencing this discussion.

**Reply by Its-Kenta** [reply]
You can also simply move the discussion over to Discord tbfh.

**Reply by RaycatWhoDat** [reply]
Yes, that would be the easier way... bu I know some of the discussion makers don't want to be on Discord for one reason or another.

**Comment by bentxt** [comment]
> Big language features missing:

Great list, may the stuff you could put under a  roadmap.txt ?

> Allocator system for manual memory management is not finished.

This is one that is indead essential, could you say a bit more about it?
